### DISCUSSION TITLE
A Windows 98 SE system for macOS (with printing and file transfer)

### DISCUSSION THREAD

**Discussion by emendelson** [joncampbell123/dosbox-x#4424] [Show and tell]
With the help of generous people here and on the Vogons forum, I've put together an AppleScript application that runs Windows 98 SE under DOSBox-X. Because Windows is copyrighted software, you'll need to unlock the application by dropping on it the .iso file on which Microsoft distributed the Windows 98 SE installation CD. Once the system is unlocked, it runs a basic installation of Windows 98 SE, with these extra features:

You can drop a file on the application icon, and it will be copied to the Windows 98 SE desktop. In Windows 98 SE, you can drop a file in a folder named "Files for macOS Host" and they will be copied to your macOS desktop when you close down Windows. 

Three "printers" are installed: one prints to the default macOS printer. Another creates an arbitrarily-named PDF on the macOS desktop. A third lets you select a printer from a list of installed macOS printers.

I've installed WinSCP (more about this in a moment), FileZilla, Netscape 2.0, the Windows Script Host, and TweakUI. You can remove any that you don't want.

If you want, you can use a macOS FTP server to exchange files with WinSCP; I use the free "Quick FTP Server" from the App Store. The key fact is to use port 2121, not 21.

When you start the app, you'll need to press a key and click the mouse before you can use the mouse cursor.

This is a first try. Please let me know what's wrong or what can be improved. The app is notarized, but the notarization may break after you use it, and you'll have to open a terminal, type the command `xattr -rc ` followed by a space, and drag the app into the terminal and press return before you can use it again. But maybe you won't need to.

I'll post a web page about this setup later, and will follow it with a Windows 10/11 version in September. Meanwhile, here's the download:

https://mendelson.org/Windows98SE.zip

**Comment by Torinde** [comment]
I'll be interested to test the Windows version and how it'll deal with various codepages (for printing, file contents, file names) in both directions.

**Comment by emendelson** [comment]
@Torinde - It's purely US English. It prints by using Ghostscript to create a PDF from the output created in Windows 98 SE for an AppleWriter IINTX, and then printing or saving the PDF. I would guess that the printing would be accurate, because Ghostscript knows how to convert the PostScript symbol sets into unicode. But that's only a guess. 

Of course, if you send a file to the Windows 98 system, it can only do what the original Windows 98 could have done with the file.

You can easily replace the hard disk image with one that contains a Windows 98 SE system in another language. I'll post instructions for doing that.

**Comment by rderooy** [comment]
One thing you can try, is the adobe postscript 3 driver. Supposedly it offers superior output and better PDF conversion, as shown here for Win 3.1x: https://github.com/dosbox-staging/dosbox-staging/wiki/Windows-3.1x#adobe-postscript-3-printer-driver

**Reply by rderooy** [reply]
Also, did the Win9x adobe acroread perhaps already provide a print-to-pdf printer option? In that case you don't even need to do the conversion on the host to PDF.

**Comment by emendelson** [comment]
@rderooy - That PostScript 3 driver was for Windows 3.x, but this reminded me that there was a PostScript 4.3.1 for Windows 9x, and I've now replaced the printer drivers in my setup with that driver, and uploaded a new version.

Also, as far as I remember, AcroRead never installed a Print-to-PDF driver; you had to buy the full Acrobat package. However, I've now added to the "Installers" folder on my system the free PDF Creator package that provides a print-to-PDF driver; and I've also added AcroRead. I haven't installed these already because I think the user should be able to choose what to install beyond the essentials that I installed in the base system (TweakUI, Windows Script Host. and WinSCP for FTP in and out of the system).

**Comment by emendelson** [comment]
Download page and description here:

https://mendelson.org/windows98se.html

**Comment by emendelson** [comment]
Updated today with a new menu option to add a CD-image so that it it's accessible from My Computer in Win98. Also, installed the official Microsoft updates (NO third-party updates that always cause trouble).

I'm curious about which `output=` setting is fastest. I'm using `openglnb` but I could be wrong. Any advice will be welcome.

**Comment by Lewiscowles1986** [comment]
I'm curious if you managed to get CD image mounting working when in Windows 98

**Reply by Lewiscowles1986** [reply]
Solved my ask

```
/Applications/dosbox-x.app/Contents/MacOS/dosbox-x -c 'IMGMOUNT 2 ~/Desktop/Games/WIN98/2GB-win98.img -fs none -size 512,63,64,1023' -c 'IMGMOUNT D: ~/Desktop/Games/WIN98/VB6EE.ISO -t iso' -c 'boot -l C' -c 'exit' -set memsize=512 -set vmemsize=32 -set cputype=pentium_iii -set fpu=true -set core=dynamic -set cycles=max
```

I'll probably do a fresh re-install with 8GB at some stage, but the executable you advertise / offer, although interesting, does too much I'd like to understand for me to leave this unsolved.

**Reply by emendelson** [reply]
It works exactly as it says on my web page. Hold down the Option key when launching the app, and choose the menu option that lets you add a CD image.

**Reply by Lewiscowles1986** [reply]
That app, packages dosbox-x in a fairly inscruitible way, not helping users to know how to self-serve. While I am sure it has a place, it is not how this should be understood; certainly not via the discussions on a source-available / open source project. It's very cool. I'm glad you made it and have shared it. But I like to understand how the sausage is made.

**Comment by emendelson** [comment]
Hmm.. I made it a run-only script in order to make it difficult to run it without a copy of the .ISO that proves that the user owns Windows 98 SE. But that may be excessive security for the moment. I may remove the run-only part of it to make it fully user-manageable.

**Reply by Lewiscowles1986** [reply]
the literal commands to dosbox-x. I Didn't see the source code listed. I Thought the main script was a compiled applescript library

I wouldn't worry about opensourcing your tool. I am interested in the lower-level (how can this be done without the wrapper). Again, not to take away from your very polished, excellent tool

**Reply by Lewiscowles1986** [reply]
And again, what worked for me, to get a CDROM was simply to mount it before issuing the boot c command as seen in https://github.com/joncampbell123/dosbox-x/discussions/4424#discussioncomment-13251619

**Comment by emendelson** [comment]
What puzzles me is that I can't seem to double-click on anything until I release the mouse and then click to lock it again. I've been experimenting with various settings, but can't make this work. - Also (not yet uploaded) I've updated DOSBox-X to the latest code, which seems to avoid the DOSBox-X crashes in the older version.

**Reply by emendelson** [reply]
> What puzzles me is that I can't seem to double-click on anything until I release the mouse and then click to lock it again. I've been experimenting with various settings, but can't make this work. 

Turns out this is because I had turbo mode turned on. I've now uploaded a new version that doesn't use turbo mode.

**Comment by emendelson** [comment]
And I had forgotten how this all works. What happens is this: when you add a CD image, the AppleScript adds a `mount` line to a file named `temporary.conf` in `~/Library/Application Support/Windows98SE` and uses that conf file to launch DOSBox-X.
